### PR TITLE
Raise an error if batch transfer hooks are overridden with IPUAccelerator

### LINKED ISSUE #13916: Warn IPU users that batch transfering hooks will not run
if that's the case, we need to do a follow-up for users to let them know that some of the hooks won't work with IPUs, like `transfer_batch_to_device`, ...

_Originally posted by @rohitgr7 in https://github.com/Lightning-AI/lightning/pull/13880#discussion_r931989445_

cc @borda

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #13916

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @justusschock @akihironitta

### CODE REVIEW

File: tests/tests_pytorch/trainer/test_config_validator.py [hunk 1]
```diff
@@ -192,3 +194,29 @@ def setup(self, pl_module, trainer):
...
+    elif trainer_kwargs.get("accelerator") == "ipu":
+        match_pattern = rf"Overriding `{hook}` is not .* with IPUs"
+        monkeypatch.setattr(pl.accelerators.ipu.IPUAccelerator, "is_available", lambda _: True)
+        monkeypatch.setattr(pl.strategies.ipu, "_IPU_AVAILABLE", lambda: True)
+
+    def custom_method(self, batch, *_, **__):
+        batch = batch + 1
+        return batch
+
+    trainer = Trainer(default_root_dir=tmpdir, max_steps=7, **trainer_kwargs)
```
Comment (awaelchli) [inline tests/tests_pytorch/trainer/test_config_validator.py:216]:
```suggestion
    trainer = Trainer(default_root_dir=tmpdir, **trainer_kwargs)
```

File: docs/source-pytorch/accelerators/gpu_intermediate.rst [hunk 2]
```diff
@@ -47,8 +47,9 @@ after which the root node will aggregate the results.
     :doc:`Manual Optimization <../model/manual_optimization>` with DP. Use DDP which is more stable and at least 3x faster.
 
 .. warning:: DP only supports scattering and gathering primitive collections of tensors like lists, dicts, etc.
-    Therefore the :meth:`~pytorch_lightning.core.hooks.ModelHooks.transfer_batch_to_device` hook does not apply in
-    this mode and if you have overridden it, it will not be called.
+    Therefore the hooks :meth:`~pytorch_lightning.core.hooks.ModelHooks.on_before_batch_transfer`,
+    :meth:`~pytorch_lightning.core.hooks.ModelHooks.transfer_batch_to_device`, :meth:`~pytorch_lightning.core.hooks.ModelHooks.on_after_batch_transfer`
+    does not apply in this mode and if you have overridden it, it will raise an exception.
```
Comment (awaelchli) [inline docs/source-pytorch/accelerators/gpu_intermediate.rst:52]:
```suggestion
    Therefore the hooks :meth:`~pytorch_lightning.core.hooks.ModelHooks.on_before_batch_transfer`,
    :meth:`~pytorch_lightning.core.hooks.ModelHooks.transfer_batch_to_device` and :meth:`~pytorch_lightning.core.hooks.ModelHooks.on_after_batch_transfer`
    do not apply in this mode and if you have overridden any of them, an exception will be raised.
```
changing to the plural also requires some grammatical changes.

File: docs/source-pytorch/accelerators/ipu_basic.rst [hunk 3]
```diff
@@ -67,3 +67,6 @@ Please see the `MNIST example <https://github.com/Lightning-AI/lightning/blob/ma
 * Since the step functions are traced, branching logic or any form of primitive values are traced into constants. Be mindful as this could lead to errors in your custom code.
 * Clipping gradients is not supported.
 * It is not possible to use :class:`torch.utils.data.BatchSampler` in your dataloaders if you are using multiple IPUs.
+* IPUs handles the data transfer to the device on the host, hence the hooks :meth:`~pytorch_lightning.core.hooks.ModelHooks.on_before_batch_transfer`,
+  :meth:`~pytorch_lightning.core.hooks.ModelHooks.transfer_batch_to_device`, :meth:`~pytorch_lightning.core.hooks.ModelHooks.on_after_batch_transfer`
+  does not apply here and if you have overridden it, it will raise an exception.
```
Comment (awaelchli) [inline docs/source-pytorch/accelerators/ipu_basic.rst:72]:
```suggestion
* IPUs handles the data transfer to the device on the host, hence the hooks :meth:`~pytorch_lightning.core.hooks.ModelHooks.on_before_batch_transfer`,
  :meth:`~pytorch_lightning.core.hooks.ModelHooks.transfer_batch_to_device` and :meth:`~pytorch_lightning.core.hooks.ModelHooks.on_after_batch_transfer`
  do not apply here and if you have overridden any of them, an exception will be raised.
```

**Review by awaelchli — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

File: src/pytorch_lightning/trainer/configuration_validator.py [hunk 4]
```diff
@@ -158,6 +159,10 @@ def __verify_dp_batch_transfer_support(trainer: "pl.Trainer", model: "pl.Lightni
             is_overridden(hook, datahook_selector.model) or is_overridden(hook, datahook_selector.datamodule)
         ):
             raise MisconfigurationException(f"Overriding `{hook}` is not supported in DP mode.")
+        if isinstance(trainer.accelerator, IPUAccelerator) and (
+            is_overridden(hook, datahook_selector.model) or is_overridden(hook, datahook_selector.datamodule)
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/configuration_validator.py:163]: You reused this method, but it's meant as a DP warning.

I suggest you change the name to something like `__verify_batch_transfer_support` and move the TODO comment right beside the DP check

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/trainer/configuration_validator.py (hunk 4 above)
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/configuration_validator.py:164]: done!
